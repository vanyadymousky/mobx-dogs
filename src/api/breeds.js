import { prop } from 'ramda';
import { ajax } from 'rxjs/ajax';
import { map } from 'rxjs/operators';

const mapResponse = map(prop('message')); // eslint-disable-line ramda/map-simplification

export const getBreeds = () =>
  ajax.getJSON('https://dog.ceo/api/breeds/list/all').pipe(mapResponse);

export const getPhotos = (breed, subbreed) => {
  const breedRequest = subbreed ? `${breed}/${subbreed}` : breed;
  return ajax.getJSON(`https://dog.ceo/api/breed/${breedRequest}/images`).pipe(mapResponse);
};
