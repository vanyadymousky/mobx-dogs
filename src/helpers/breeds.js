import { compose, length, values } from 'ramda';

export const getCount = compose(length, values);
