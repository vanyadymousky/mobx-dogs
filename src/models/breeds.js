import { observable, action } from 'mobx';
import { toPairs } from 'ramda';
import { catchError, tap } from 'rxjs/operators';
import { getBreeds, getPhotos } from '../api/breeds';
import { getCount } from '../helpers/breeds';

export const createBreed = (name, subbreeds) => observable({
  name,
  subbreeds,
});

export const breedsStore = observable({
  items: [], // collection of breeds
  images: {},
  inProgress: false,

  get count() {
    return getCount(this.items);
  },

  startProgress() {
    this.inProgress = true;
  },

  stopProgress() {
    this.inProgress = false;
  },

  addBreed(breed) {
    this.items.push(breed);
  },

  fetch() {
    this.startProgress();
    return getBreeds().pipe(
      tap(this.stopProgress),
      catchError(this.stopProgress),
    ).subscribe((breeds) => {
      this.items = toPairs(breeds).map(breed => createBreed(...breed));
    });
  },

  fetchImages(breed) {
    this.startProgress();
    return getPhotos(breed).pipe(
      tap(this.stopProgress),
      catchError(this.stopProgress),
    ).subscribe((photos) => {
      this.images[breed] = photos;
    });
  },
}, {
  addBreed: action,
  fetch: action,
  fetchImages: action,
  startProgress: action.bound,
  stopProgress: action.bound,
});

// autorun(() => console.log(breedsStore.items));
