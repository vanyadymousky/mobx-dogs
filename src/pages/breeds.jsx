import React from 'react';
import { map, isEmpty } from 'ramda';
import { observer } from 'mobx-react';
import { breedsStore } from 'models/breeds';
import { IfElse } from '../components/helpers/if-else';
import { StateLink } from '../components/state-link/state-link';

const renderBreeds = map(({ name }) => (
  <li key={name}>
    <StateLink to={`breed/${name}`}>
      {name}
    </StateLink>
  </li>
));

@observer
class Breeds extends React.Component {
  componentDidMount() {
    breedsStore.fetch();
  }

  render() {
    const { inProgress, items, count } = breedsStore;

    return (
      <div>
        <h1>Hey hey breeds!</h1>
        <IfElse condition={inProgress}>
          <p>Loading madafaka....</p>
          <h4>{count} breeds loaded</h4>
        </IfElse>
        <ul>
          {!isEmpty(items) && renderBreeds(items)}
        </ul>
      </div>
    );
  }
}

export default Breeds;
