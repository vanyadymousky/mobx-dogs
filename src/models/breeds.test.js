import { toJS } from 'mobx';
import { breedsStore, createBreed } from './breeds';

describe('breeds store', () => {
  test('should be default from scratch', () => {
    expect(toJS(breedsStore)).toMatchSnapshot();
  });

  test('should be in progress', () => {
    breedsStore.startProgress();
    expect(breedsStore.inProgress).toBe(true);
  });

  test('should add items', () => {
    breedsStore.addBreed(createBreed('testme', []));
    expect(toJS(breedsStore)).toMatchSnapshot();
  });
});

