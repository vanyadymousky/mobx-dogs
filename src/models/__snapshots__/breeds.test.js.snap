// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`breeds store should add items 1`] = `
Object {
  "images": Object {},
  "inProgress": true,
  "items": Array [
    Object {
      "name": "testme",
      "subbreeds": Array [],
    },
  ],
}
`;

exports[`breeds store should be default from scratch 1`] = `
Object {
  "images": Object {},
  "inProgress": false,
  "items": Array [],
}
`;
